### PR TITLE
Config and custom queries for JS/JSX/TS/TSX

### PR DESCRIPTION
**Pull request by rosingrind**
I've taken queries from `helix` main, processed and tested them a bit on JS/JSX/TS/TSX files. All seems good:

Screenshot 2023-09-03 at 12 11 48 PM

But I have some questions:

- [x] is `</` part of `</div>` or any other closing tag highlighted grey because of semantic highlighting or I did something wrong?
- [x] how can I add the license, should I use other `kak-tree-sitter` bundled queries as an example or you'll do it yourself?
- there is a caveat when using `javascript` parser for `kts_lang == jsx`: https://github.com/tree-sitter/tree-sitter-javascript/issues/244
  - `tree-sitter-javascript` is written for `javascript` languageId
  - so you need to modify source files on `-c` stage of `ktsctl -fci jsx` run
  - I've come up with a simple shell command to replace all `javacscript` mentions in parser code: `grep -rl 'javascript' . | xargs sed -i '' -e 's/javascript/jsx/g'`
    - [ ] this is a manual and pretty dirty hack, but it is what it is - is it ok for now or you do have any other ideas? I couldn't inject this command right to `language.jsx.grammar.compile` because [piping](https://doc.rust-lang.org/std/process/struct.Command.html) mechanisms of rust will require editing `ktsctl` sources
  - if you did prepared your sources for `jsx`, everything will work and `jsx` language will be recognized
  - info: `tsx` does not have such problems because of split `tsx` and `typescript` parsers in their repo

Queries are merged according to doc: https://github.com/helix-editor/helix/tree/master/runtime/queries/ecma#inheritance-model-for-ecma-based-languages

### CODE REVIEW

File: config.toml [hunk 1]
```diff
@@ -488,7 +488,21 @@ pin = "dbd248fdfa680373d94fbc10094a160aafa0f7a7"
...
+compile = "cc"
+compile_args = ["-c", "-fpic", "../scanner.c", "../parser.c", "-I", ".."]
+compile_flags = ["-O3"]
+link = "cc"
+link_args = ["-shared", "-fpic", "scanner.o", "parser.o", "-o", "javascript.so"]
+link_flags = ["-O3"]
+
+[language.javascript.queries]
+url = "https://github.com/phaazon/kak-tree-sitter"
+pin = "*"
```
Comment (rosingrind) [inline config.toml:504]: every `pin` entry for queries needs to be replaced

**Comment by rosingrind** [comment]
By the way, if you plan to merge this, I found it great to switch `kts_lang` automatically with the following hook:

```kakounescript
hook global BufSetOption kts_lang=(javascript|typescript) %{
	eval %sh{
		case $kak_bufname in
			(*\.jsx) echo "set-option buffer kts_lang 'jsx'";;
			(*\.tsx) echo "set-option buffer kts_lang 'tsx'";;
		esac
	}
}
```

As far as I understand, kakoune itself doesn't differ js and jsx filetype (same for ts and tsx), so does `kak-tree-sitter`. I've come up with this hack, but I'm not sure if this is right approach as it may clash with plugin's initialization. If it's an ok approach, this hook snippet probably should be referenced somewhere in wiki

**Comment by phaazon** [comment]
> how can I add the license, should I use other kak-tree-sitter bundled queries as an example or you'll do it yourself?

If you are just cloning remote repositories, I think it’s all good, the licenses should be copied over. If you are vendoring them in `kak-tree-sitter/runtime/queries`, then include a `README.md` and the licenses if you copied any files (you have plenty of examples in that directory already).

> is </ part of </div> or any other closing tag highlighted grey because of semantic highlighting or I did something wrong?

Try a `rmhl window/semantic-tokens`.

> By the way, if you plan to merge this […]

Sure, that’s the idea!

**Comment by rosingrind** [comment]
`rmhl window/semantic-tokens` didn't change anything for me. I didn't enable `lsp_semantic_tokens` from `kak-lsp` and I don't have a direct `window/semantic-tokens` entry btw, I was just assuming that kts have some sort of semantic highlighting that makes tags look like this :) I think it's probably alright because if jsx tag parsing has failed, there will be no hl at all - I'm just making sure that `</` is not dropped from parsing

**Comment by rosingrind** [comment]
I think I found a proof it parses correctly: pressing `m` on any part of closing tag (e.g. `<` `/` or `>`) selects the whole tag. It's just a highlighting thing, and it may be done as needed via #78

**Comment by rosingrind** [comment]
It's the same for `.jsx` and `.tsx` files. I don't really understand why highlighting works like that because there are `jsx_opening_element`, `jsx_closing_element` and `jsx_self_closing_element` entries in jsx queries (and they are all of `@punctuation.bracket`, so `<` should be colored like `</`?), but if there's no issues with it - I think it's better to work on these parts in #78

**Comment by rosingrind** [comment]
Ok, the queries are correct and everything works as expected! I've done more testing and found that if I remove these lines

https://github.com/phaazon/kak-tree-sitter/blob/23117af3692a99a01d31c6b5328065059a86d8ec/runtime/queries/jsx/highlights.scm#L290-L294

then the sample source file gets highlighted like this

Screenshot 2023-09-03 at 11 48 32 PM

which means that everything is ok, it's just highlighting mechanisms that works like that (you can change `@punctuation.bracket` to `@function.macro` and get `kts_mauve` colored `</`, same as the `function` keyword on line 5 of sample file)

**Comment by rosingrind** [comment]
PR is rebased onto master, I'll summarize some additional info:

- `javascript`, `typescript` and `tsx` works out of the box, you can run `ktsctl -fci *` and get it working
- `jsx`, because of non-split architecture, is installed as follows:
  - `ktsctl -f jsx`
  - replace all `javascript` occurrences in fetched dir `tree-sitter-javascript/src` with `jsx` (`grep -rl 'javascript' . | xargs sed -i '' -e 's/javascript/jsx/g'`)
  - `ktsctl -ci jsx`

Useful hook:

> ```kakounescript
> hook global BufSetOption kts_lang=(javascript|typescript) %{
> 	eval %sh{
> 		case $kak_bufname in
> 			(*\.jsx) echo "set-option buffer kts_lang 'jsx'";;
> 			(*\.tsx) echo "set-option buffer kts_lang 'tsx'";;
> 		esac
> 	}
> }
> ```

**Comment by phaazon** [comment]
Okay, but that’s probably too much thing to do for people. Isn’t there a way to simplify the process?

**Comment by rosingrind** [comment]
For now I see these solutions:

- contribute to `tree-sitter-javascript` and split `jsx` out
- add `pre-build` routine to `ktsctl` which will call sh file or the replacement commands directly
- make a fork of `tree-sitter-javascript` with patching languageId and regeneration of sources via CI
- drop JS and JSX support for now until one of previous solutions is done

Only the first one seems like a proper one, but I didn't have much time lately to do this

**Comment by phaazon** [comment]
Well, I’m okay to support and patch more things as long as it’s not bending the rest / adding language specificities to generic code paths. So basically, if someone wants Javascript and JSX, they have to go the JSX way. We can probably add that in the wiki.

> add `pre-build` routine

We can probably discuss that but I’m not entirely sure about this approach. It feels like a flood gate that could be wide opened.

File: config.toml (hunk 1 above)
Comment (phaazon) [reply config.toml:504]: Is this still up to date?

File: config.toml [hunk 2]
```diff
@@ -514,7 +528,21 @@ path = "runtime/queries/json"
...
+compile = "cc"
+compile_args = ["-c", "-fpic", "../scanner.c", "../parser.c", "-I", ".."]
+compile_flags = ["-O3"]
+link = "cc"
+link_args = ["-shared", "-fpic", "scanner.o", "parser.o", "-o", "jsx.so"]
+link_flags = ["-O3"]
+
+[language.jsx.queries]
+url = "https://github.com/phaazon/kak-tree-sitter"
+pin = "*"
```
Comment (phaazon) [inline config.toml:544]: Ditto.

File: config.toml [hunk 3]
```diff
@@ -848,13 +876,41 @@ path = "runtime/queries/toml"
...
+compile = "cc"
+compile_args = ["-c", "-fpic", "../scanner.c", "../parser.c", "-I", ".."]
+compile_flags = ["-O3"]
+link = "cc"
+link_args = ["-shared", "-fpic", "scanner.o", "parser.o", "-o", "tsx.so"]
+link_flags = ["-O3"]
+
+[language.tsx.queries]
+url = "https://github.com/phaazon/kak-tree-sitter"
+pin = "*"
```
Comment (phaazon) [inline config.toml:892]: Ditto.

**Review by phaazon — COMMENTED** [review]
I think we should be merging that as soon as the `pin` question is resolved.

**Comment by rosingrind** [comment]
Removed all `pin = "*"` entries and rebased onto master, all should be up-to-date

**Review by phaazon — APPROVED** [review]

**Comment by phaazon** [comment]
Do people still need the `typescript` grammar for typescript files? I’m not sure I understand fully the hook. If the file is tagged as `typescript`, then it stays `typescript`? The hook should be on `javascript` only then, no?

**Comment by rosingrind** [comment]
I think yes, because tree-sitter parser for typescript and tsx differs. I you won't use hook, you will run raw typescript parser with tsx queries on `.tsx` file, which in my case gave me incorrect highlighting. Queries are basically the same except [jt]sx queries contain `<tag>` entries.

- typescript and tsx are different parsers separated in their repo
- javascript is the unified parser for javascript and jsx files

**Comment by phaazon** [comment]
Okay, then the hook is wrong, it should set `kts_lang` to `jsx`, not `tsx`. I’ll update the wiki. Thanks for replying that fast!

**Comment by rosingrind** [comment]
You should be able to use tsx grammar on typescript files, but I haven't checked that properly - it worked on my side, but as these are 2 separated parsers, some issues are possible

**Comment by rosingrind** [comment]
> it should set `kts_lang` to `jsx`, not `tsx`

- jsx for `*.jsx` files
- tsx for `*.tsx` files

**Comment by phaazon** [comment]
https://github.com/phaazon/kak-tree-sitter/wiki/JSX

**Comment by rosingrind** [comment]
Thanks for the wiki page! Everything seems ok

**Pull request merged**